### PR TITLE
HHH-13846 Fix TemplateRenderer's param index parsing bug

### PR DESCRIPTION
**Pull request by NathanQingyangXu**
https://hibernate.atlassian.net/browse/HHH-13846

Initially I thought this was an easy fix, but ...

### CODE REVIEW

File: hibernate-core/src/main/java/org/hibernate/dialect/function/TemplateRenderer.java [hunk 1]
```diff
@@ -74,7 +76,7 @@ public TemplateRenderer(String template) {
 			chunkList.add( chunk.toString() );
 		}
 
-		chunks = chunkList.toArray( new String[chunkList.size()] );
+		chunks = chunkList.toArray( new String[0] );
```
Comment (NathanQingyangXu) [inline hibernate-core/src/main/java/org/hibernate/dialect/function/TemplateRenderer.java]: As per https://stackoverflow.com/questions/174093/toarraynew-myclass0-or-toarraynew-myclassmylist-size, seems empty array approach performs better, let alone its better maintainability.

File: hibernate-core/src/main/java/org/hibernate/dialect/function/TemplateRenderer.java [hunk 2]
```diff
@@ -61,8 +61,10 @@ public TemplateRenderer(String template) {
 					}
 				}
 
-				paramList.add( Integer.valueOf( index.toString() ) );
-				index.delete( 0, index.length() );
+				if ( index.length() > 0 ) {
+					paramList.add( Integer.valueOf( index.toString() ) );
+				}
+				index.setLength( 0 );
```
Comment (NathanQingyangXu) [inline hibernate-core/src/main/java/org/hibernate/dialect/function/TemplateRenderer.java]: I think 'setLength()' is a little bit better in terms of both performance and code readability. See https://stackoverflow.com/questions/5192512/how-can-i-clear-or-empty-a-stringbuilder

**Comment by gavinking** [comment]
@NathanQingyangXu but does this really completely fix the problem? With this change don't you get a mismatch between the length of `chunkList` and the length of the `paramList`?

**Comment by gavinking** [comment]
TBH, I'm not even sure I think this is a bug. If there's a strong need to allow literal `?` in patterns, I would introduce a well-defined syntax for it (`??` perhaps) but really I would say the main problem here is just the exception:

    java.lang.NumberFormatException: For input string: ""

If we caught that and produced a better message, I think we would be good...

**Comment by NathanQingyangXu** [comment]
> @NathanQingyangXu but does this really completely fix the problem? With this change don't you get a mismatch between the length of `chunkList` and the length of the `paramList`?

Good catch! Yeah, it is much more complicated than I've thought initially. If '?' doesn't follow digit, it should be included in some chunk. I updated the code to make that happen, but I feel it is becoming more and more complicated...

**Comment by NathanQingyangXu** [comment]
> TBH, I'm not even sure I think this is a bug. If there's a strong need to allow literal `?` in patterns, I would introduce a well-defined syntax for it (`??` perhaps) but really I would say the main problem here is just the exception:
> 
> ```
> java.lang.NumberFormatException: For input string: ""
> ```
> 
> If we caught that and produced a better message, I think we would be good...

Good point! Yeah, this seems a really edgy case and it seems not worthwhile to fix it ending up with making the code complicated (there is no foolproof solution following the existing direction).

I enriched the testing case but I am afraid it might not be worthwhile. @sebersole , what do you think?

File: hibernate-core/src/main/java/org/hibernate/dialect/function/TemplateRenderer.java [hunk 3]
```diff
@@ -47,22 +47,28 @@ public TemplateRenderer(String template) {
...
 						break;
 					}
 				}
 
-				paramList.add( Integer.valueOf( index.toString() ) );
-				index.delete( 0, index.length() );
+				if ( index.length() > 0 ) {
+					paramList.add( Integer.valueOf( index.toString() ) );
+					index.setLength( 0 );
+				}
```
Comment (NathanQingyangXu) [inline hibernate-core/src/main/java/org/hibernate/dialect/function/TemplateRenderer.java]: The basic idea is simple. If '?' is not followed by digit(s), it should be teated as normal chunk char. However, the code seems unnecessarily complicated and we might refactor the logic by simple RE processing instead.

File: hibernate-core/src/test/java/org/hibernate/test/dialect/function/TemplateRendererTest.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+				+ "    REPLACE( "
+				+ "        TRANSLATE( "
+				+ "            LOWER(?1), "
+				+ "            'áàãâäéèêëíìîïóòõôöúùûüç', "
+				+ "            'aaaaaeeeeiiiiooooouuuuc' "
+				+ "        ), "
+				+ "        ' ', "
+				+ "        '' "
+				+ "    ), "
+				+ "    '[\\u2000-\\u206F\\u2E00-\\u2E7F\\\\''!\"#$%&()*+,\\-.\\/:;<=>?@\\[\\]^_`{|}~]', " // -- '?'' causes the problem
```
Comment (NathanQingyangXu) [inline hibernate-core/src/test/java/org/hibernate/test/dialect/function/TemplateRendererTest.java]: if the above '?' is followed by digit, it should not be treated as param, so I doubt whether the fixing is worthwhile in the first place.

**Comment by NathanQingyangXu** [comment]
> TBH, I'm not even sure I think this is a bug. If there's a strong need to allow literal `?` in patterns, I would introduce a well-defined syntax for it (`??` perhaps) but really I would say the main problem here is just the exception:
> 
> ```
> java.lang.NumberFormatException: For input string: ""
> ```
> 
> If we caught that and produced a better message, I think we would be good...

Yeah, I have to admit you have a good point. I found I am stuck somewhere. @sebersole is providing more meaningful exception in constructor is acceptable here?

**Comment by NathanQingyangXu** [comment]
This bugfix is more complicated than I originally thought. For that reason, I decided to close the PR and create a new one when I am ready. Thanks for excellent comments from @gavinking

**Comment by mhagnumdw** [comment]
@NathanQingyangXu ,  I think this can solve:

```java
public TemplateRenderer(String template) {
    this.template = template;

    final List<String> chunkList = new ArrayList<String>();
    final List<Integer> paramList = new ArrayList<Integer>();
    final StringBuilder chunk = new StringBuilder( 10 );
    final StringBuilder index = new StringBuilder( 2 );

    int i = 0;
    final int len = template.length();
    while ( i < len ) {
        char c = template.charAt( i );
        if ( c == '?' ) { // Is it a parameter or literal? We will see ...

            while ( ++i < template.length() ) {
                c = template.charAt( i );
                if ( Character.isDigit( c ) ) {
                    index.append( c );
                }
                else {
                    if (index.length() > 0) { // the '?' represent an argument
                        chunkList.add( chunk.toString() );
                        chunk.delete( 0, chunk.length() );
                        
                        paramList.add( Integer.valueOf( index.toString() ) );
                        index.delete( 0, index.length() );
                    } else { // the '?' represent a literal
                        chunk.append( '?' );
                    }
                    chunk.append( c );
                    break;
                }
            }
            
        }
        else {
            chunk.append( c );
        }
        i++;
    }

    if ( chunk.length() > 0 ) {
        chunkList.add( chunk.toString() );
    }

    chunks = chunkList.toArray( new String[chunkList.size()] );
    paramIndexes = new int[paramList.size()];
    for ( i = 0; i < paramIndexes.length; ++i ) {
        paramIndexes[i] = paramList.get( i );
    }
}

```

Diff (before / after)

image

- I believe that if the '?' is followed by a digit should always be treated as a parameter
- I tested it for the test case you created and it worked as expected
- Unfortunately my resources are limited at the moment and therefore I have no way to run the Hibernate tests

What do you think?

**Comment by NathanQingyangXu** [comment]
Thanks for following up. TBH, I think the second comment from Gavin King
makes sense and I don't think there is fool-proof fix following the current
direction. For that reason, I closed my PR. Is the issue really serious
without bypassing?
As to the code change, I think it is becoming unnecessarily complicated.
For example, if the template ends with '?', seems it will be ignored. I
made similar code changes in my closed PR, but I doubt it is correct
direction. Suppose there is a digit after the '?', it seems we cannot say
it is intended param. Maybe a more complete syntax design is needed.
That being said, you are welcome to create PR on top of my code. But I
would think more on the best solution.

**Pull request closed without merging**